File: app/schemas/wyckoff.py
"""威科夫方法论分析请求/响应 Schema。"""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class TradingRangeOut(BaseModel):
    kind: Literal["accumulation", "distribution"]
    support: float
    resistance: float
    start_time: str
    end_time: str


class WyckoffEventOut(BaseModel):
    code: str          # SC / AR / ST / SPRING / SOS / UT / UTAD / SOW / LPS / LPSY / PS / PSY / BC / TEST / BU
    name: str          # 中文名
    time: str
    price: float
    volume_ratio: float
    phase: str         # A / B / C / D / E
    description: str


class LawOut(BaseModel):
    key: Literal["supply_demand", "cause_effect", "effort_result"]
    name: str
    verdict: str
    detail: str


class PhaseOut(BaseModel):
    stage: str         # accumulation / markup / distribution / markdown / undetermined
    stage_label: str
    phase: str         # A–E
    phase_label: str
    breakout: Literal["up", "down", "none"]


class RecommendationOut(BaseModel):
    action: str        # accumulate / buy / hold / reduce / avoid / watch
    action_label: str
    bias: Literal["bullish", "bearish", "neutral"]
    reasons: list[str]
    caveats: list[str]


class CandleOut(BaseModel):
    time: str
    open: float
    high: float
    low: float
    close: float
    volume: float


class WyckoffAnalysisResponse(BaseModel):
    symbol: str
    bars_count: int
    context: str
    candles: list[CandleOut]
    trading_range: Optional[TradingRangeOut]
    events: list[WyckoffEventOut]
    phase: Optional[PhaseOut]
    laws: list[LawOut]
    stage_label: str
    position_desc: str
    summary: str
    recommendation: Optional[RecommendationOut] = None

File: frontend/lib/api/wyckoff.ts
import apiClient from './client'

export interface Candle {
  time: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface TradingRange {
  kind: 'accumulation' | 'distribution'
  support: number
  resistance: number
  start_time: string
  end_time: string
}

export interface WyckoffEvent {
  code: string
  name: string
  time: string
  price: number
  volume_ratio: number
  phase: string
  description: string
}

export interface WyckoffLaw {
  key: 'supply_demand' | 'cause_effect' | 'effort_result'
  name: string
  verdict: string
  detail: string
}

export interface WyckoffPhase {
  stage: string
  stage_label: string
  phase: string
  phase_label: string
  breakout: 'up' | 'down' | 'none'
}

export interface Recommendation {
  action: string
  action_label: string
  bias: 'bullish' | 'bearish' | 'neutral'
  reasons: string[]
  caveats: string[]
}

export interface WyckoffAnalysisResult {
  symbol: string
  bars_count: number
  context: string
  candles: Candle[]
  trading_range: TradingRange | null
  events: WyckoffEvent[]
  phase: WyckoffPhase | null
  laws: WyckoffLaw[]
  stage_label: string
  position_desc: string
  summary: string
  recommendation: Recommendation | null
}

export async function fetchWyckoffAnalysis(
  symbol: string,
  startDate: string,
  endDate: string,
  freq = 'daily',
): Promise<WyckoffAnalysisResult> {
  const res = await apiClient.get<WyckoffAnalysisResult>('/api/v1/wyckoff/analysis', {
    params: { symbol, start_date: startDate, end_date: endDate, freq },
  })
  return res.data
}
